Initialization, CTAD, and `[[no_unique_address]]` improvements

brycelelbach opened the pull request:
Fixes #58.

crtrott committed:
Fixed single header generation

brycelelbach committed:
Refactor layout mappings to be implemented as nested template structs not
nested alias templates in preparation for layout mapping deduction guides.
Class Argument Template Deduction is in C++17 and is available in all major
compilers, while Alias Template Argument Deduction is only in C++20 and
currently only implemented in GCC.

brycelelbach committed:
Replace `_MDSPAN_USE_DEDUCTION_GUIDES` with
`_MDSPAN_USE_CLASS_TEMPLATE_ARGUMENT_DEDUCTION` (C++17) and
`_MDSPAN_USE_ALIAS_TEMPLATE_ARGUMENT_DEDUCTION` (C++20).

brycelelbach committed:
* Add deduction guides for deducing `extents` of all `dynamic_extents`.
* Add `dextents` alias and tests.

brycelelbach committed:
* Add support and codepaths for C++20's [[no_unique_address]].
* Add many more tests for `detail::__compressed_pair`'s layout and ABI.

brycelelbach committed:
Allow `mdspan`s to be constructed from a `array` of extents of a type
convertible to `size_t`; P0009R12 allows this but the implementation didn't
actually support it.

brycelelbach committed:
* Add `mdspan` deduction guides for the `array` extents constructor and
  mapping/accessor constructors. Now all constructors have deduction guides.
* Improve the `mdspan` deduction guide for the constructor that takes a
  parameter pack of dynamic indices.
* Change `layout_stride::mapping`'s constructor to take a `dextents` for the
  strides instead of an `array`. This is both cleaner, and was needed to make
  CTAD work.
* Add a `layout` type alias to all layout mappings, which is needed by
  `mdspan`'s deduction guides for its layout mapping constructors.
* Move all `mdspan` initialization and CTAD tests into `test_mdspan_ctors`.
* Add `layout_(left|right)` and `layout_stride` initialization and CTAD tests to
  `test_layout_ctors` and `test_layout_stride` respectively.
* Add a test for `mdspan` construction from a C-array.
* Add a test for `mdspan` construction from an extents `array` of type
  convertible to, but not exactly, `size_t`.

brycelelbach committed:
Fix `make_single_header` to detect includes that have spaces between the `#`
and the `include`.

brycelelbach committed:
Fix CTAD configuration macros to enable CTAD with Clang.

brycelelbach committed:
Rename `__impl_impl` to something less vague.

brycelelbach committed:
Remove the `mdspan` alias and rename `basic_mdspan` to `mdspan`.

crtrott commented:
probably needs a rebase?

crtrott commented on benchmarks/copy/copy_layout_stride.cpp:
```diff
@@ -335,7 +335,7 @@ void BM_Raw_Copy_2D(benchmark::State& state, T, size_t x, size_t y) {
   {
     // just for setup...
-    auto wrapped = stdex::mdspan<T, stdex::dynamic_extent>{buffer.get(), x * y};
+    auto wrapped = stdex::mdspan<T, stdex::extents<1>>{buffer.get(), x * y};
```
should be dextents ...

brycelelbach replied:
Fixed

brycelelbach commented:
Rebased.

brycelelbach committed:
Use `#pragma once` instead of an include guard in `config.hpp`.

brycelelbach committed:
Use an include guard instead of `#pragma once` in single file header generation.

URL includes in Godbolt appear to be expanded into the input source directly,
not through the C++ preprocessor, which will cause warnings if you use `#pragma
once` in your single file header, because most compilers warn if you have
`#pragma once` in a non-header file.

brycelelbach committed:
Add GitHub action to automatically build and publish the single header.

crtrott commented on benchmarks/stencil/cuda/stencil_3d_cuda.cu:
```diff
@@ -204,7 +206,7 @@ void BM_Raw_Cuda_Stencil_3D_right(benchmark::State& state, T, SizeX x, SizeY y,
   {
     // just for setup...
-    auto wrapped = stdex::mdspan<T, stdex::dynamic_extent>{};
+    auto wrapped = stdex::mdspan<T, stdex::extents<1>>{};
```
needs to be dextents ...

crtrott commented on benchmarks/stencil/cuda/stencil_3d_cuda.cu:
```diff
@@ -204,7 +206,7 @@ void BM_Raw_Cuda_Stencil_3D_right(benchmark::State& state, T, SizeX x, SizeY y,
   {
     // just for setup...
-    auto wrapped = stdex::mdspan<T, stdex::dynamic_extent>{};
+    auto wrapped = stdex::mdspan<T, stdex::extents<1>>{};
```
Needs to be dextents

brycelelbach replied:
Done.

crtrott commented on benchmarks/stencil/stencil_3d.cpp:
```diff
@@ -100,17 +102,17 @@ void BM_MDSpan_Stencil_3D(benchmark::State& state, MDSpan, DynSizes... dyn) {
 void BM_Raw_Stencil_3D_right(benchmark::State& state, T, SizeX x, SizeY y, SizeZ z) {
 
-  using MDSpan = stdex::mdspan<T, stdex::dynamic_extent, stdex::dynamic_extent, stdex::dynamic_extent>;  
+  using MDSpan = stdex::mdspan<T, stdex::extents<3>>;
```
needs to be dextents

brycelelbach replied:
Done.

crtrott commented on include/experimental/__p0009_bits/config.hpp:
```diff
@@ -61,13 +60,15 @@
 #endif
 
-static_assert(_MDSPAN_CPLUSPLUS >= 201102L, "MDSpan requires C++11 or later.");
+static_assert(_MDSPAN_CPLUSPLUS >= 201102L, "mdspan requires C++11 or later.");
```
lets switch this to C++14 here. No need to support 11

brycelelbach replied:
Done.

brycelelbach replied:
Done.

crtrott commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -63,87 +64,111 @@ struct layout_left_idx_conditional {
+    // serves as a pseudo deduction guide, we can get the automatic deduction
+    // guide to do the right thing.
+    MDSPAN_INLINE_FUNCTION
+    constexpr mapping(Extents const&) noexcept = delete;
```
isn't that constructor supposed to exist??

crtrott replied:
oh wait are we going via some partial specialization below?

brycelelbach replied:
Nope, I refactored my way out of this. It's no longer needed.

nliber commented on include/experimental/__p0009_bits/compressed_pair.hpp:
```diff
@@ -0,0 +1,218 @@
+#include "trait_backports.hpp"
+
+#if !defined(_MDSPAN_USE_ATTRIBUTE_NO_UNIQUE_ADDRESS)
+#  include "no_unique_address.hpp"
```
That's a weird spacing for `#  include` (in many places for this header only).

brycelelbach replied:
We do it in `config.hpp`. I'm not a fan, I was just following what `config.hpp` does.

brycelelbach commented:
Making CTAD work for the layout mappings was interesting, because they are member template classes or aliases. I changed them to not be aliases, because I wanted to avoid depending on alias template deduction, which is C++20 only and not available everywhere yet. However, this has one notable downside: [GCC currently rejects deduction guides for member template classes](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100983), so if the layout mappings are member template classes instead of aliases, we can't write deduction guide for them.

For `layout_left` and `layout_right`, the automatic deduction guides should work - they have a single template parameter which is an extents object and they have a constructor that takes a single extents object. However, we implement the layout mappings as a template specialization, and CTAD looks at the primary template for deduction. The primary template for the layout mappings was previously declared and not defined, and now just contains a `static_assert`.

We could refactor the code to not implement the mappings as a template specialization, but that seemed like a bunch of work. Without that refactoring, we need a deduction guide to make `layout_left` and `layout_right` work, but as stated earlier, I couldn't add one if they were member class templates because GCC is broken, and I didn't want to make them member alias templates because I'd have to rely on alias template argument deduction which is C++20 only and only supported by GCC. However, I was able to workaround this problem by introducing deleted constructors to the primary template.

brycelelbach commented:
> We could refactor the code to not implement the mappings as a template specialization, but that seemed like a bunch of work.

Turns out this was a lot less work than I thought, so I did it. Now the layout mappings are not implemented as partial specializations, but instead as primary templates, which makes enabling automatic deduction guides a lot simpler.

brycelelbach committed:
Refactor layout mappings so that they are implemented as primary templates not
partial specializations, which makes it easier to enable automatic deduction
guides for them.

dalg24 commented on include/experimental/__p0009_bits/depend_on_instantiation.hpp:
```diff
@@ -41,18 +41,21 @@
 */
 
-#include <experimental/mdspan>
+#pragma once
```
That's the silliest diff I have seen in a while :/
You broke GitHub

brycelelbach replied:
Not sure what you mean?

dalg24 replied:
The diff displayed is absurd because git thinks that `tests/test_ctors.cpp -> .../__p0009_bits/depend_on_instantiation.hpp`

dalg24 replied:
Your code is fine ;)

brycelelbach committed:
* Refactor `mdspan` to be implemented as a primary template instead of a partial
  specialization to simplify CTAD.
* Add a missing `mdspan(T*, extents)` constructor and deduction guide.

brycelelbach committed:
Add a link to a Godbolt environment for playing around with `mdspan` to the README.

brycelelbach committed:
Always check for the CTAD enablement macro with `defined`.

dalg24 commented on include/experimental/__p0009_bits/layout_right.hpp:
```diff
@@ -46,6 +46,7 @@
 #include "macros.hpp"
 #include "fixed_layout_impl.hpp"
 #include "trait_backports.hpp"
+#include "depend_on_instantiation.hpp"
```
Where is it used?

brycelelbach replied:
This was used in a static_assert that's been removed, it can probably be removed as well.

dalg24 replied:
You forgot to remove that guy

dalg24 replied:
Nevermind you fixed it

crtrott reviewed (changes requested):
Largely ok other than some dextents missing the d. We need to test on more compilers first, but I think this is good.

crtrott reviewed (changes requested):
Still missing a couple of "d"s

dalg24 commented on include/experimental/__p0009_bits/layout_stride.hpp:
```diff
@@ -48,6 +48,12 @@
 #include "static_array.hpp"
 #include "extents.hpp"
 #include "trait_backports.hpp"
+#include "depend_on_instantiation.hpp"
```
Where is it used?

brycelelbach replied:
This was used in a static_assert that's been removed, it can probably be removed as well.

brycelelbach committed:
Require C++14.

brycelelbach committed:
Remove the `depend_on_instantiation` facility that is no longer used.

crtrott reviewed (changes requested):
This looks good. but thanks to lack of proper testing we broke C++14 compilation already before so I can't check whether this does further damage ....

Looking into this now. 1e40f4eeab3281103 still worked, cc22d07f566 is broken.

crtrott commented:
looks like the compiler is lying about the _v stuff available??

crtrott commented:
I was able to fix the GCC 5 build issue fairly simple, since it was broken before I think we can do it as a separate PR. still having a problem with Intel compiler (17)

crtrott commented:
Intel 17 build also broke with the 1e40f4e SHA so no hold up from that either I think.

dalg24 commented on .github/workflows/single-header.yml:
```diff
@@ -0,0 +1,23 @@
+  push:
+    branches:
+      - stable
+  pull_request:
```
I think we need to remove this line

dalg24 committed:
Do not trigger single-header script on PR
